source/lex.tex

```diff
@@ -1454,64 +1473,64 @@
-are ignored when determining its value.
-\begin{example}
-The floating-point literals \tcode{1.602'176'565e-19} and \tcode{1.602176565e-19}
-have the same value.
-\end{example}
-Either the integer part or the fraction part (not both) can be omitted.
-Either the radix point or the letter \tcode{e} or \tcode{E} and
-the exponent (not both) can be omitted from a decimal floating-point literal.
-The radix point (but not the exponent) can be omitted
-from a hexadecimal floating-point literal.
-The integer part, the optional radix point, and the optional fraction part,
-form the \defn{significand} of the floating-point literal.
-In a decimal floating-point literal, the exponent, if present,
-indicates the power of 10 by which the significand is to be scaled.
-In a hexadecimal floating-point literal, the exponent
-indicates the power of 2 by which the significand is to be scaled.
-\begin{example}
-The floating-point literals \tcode{49.625} and \tcode{0xC.68p+2} have the same value.
-\end{example}
-If the scaled value is in
-the range of representable values for its type, the result is the scaled
-value if representable, else the larger or smaller representable value
-nearest the scaled value, chosen in an \impldef{choice of larger or smaller value of
-floating-point literal} manner.
-\indextext{literal!\idxcode{double}}%
-The type of a floating-point literal is \tcode{double}
 \indextext{literal!type of floating-point}%
-unless explicitly specified by a suffix.
 \indextext{literal!\idxcode{float}}%
 \indextext{suffix!\idxcode{F}}%
 \indextext{suffix!\idxcode{f}}%
-The suffixes \tcode{f} and \tcode{F} specify \tcode{float},
 \indextext{suffix!\idxcode{L}}%
 \indextext{suffix!\idxcode{l}}%
 \indextext{literal!\idxcode{long double}}%
-the suffixes \tcode{l} and \tcode{L} specify \tcode{long}
-\tcode{double}. If the scaled value is not in the range of representable
+The type of a \grammarterm{floating-point-literal} is determined by
+its \grammarterm{floating-point-suffix} as specified in \tref{lex.fcon.type}.
+\begin{simpletypetable}
+{Types of \grammarterm{floating-point-literal}{s}}
+{lex.fcon.type}
+{ll}
+\topline
+\lhdr{\grammarterm{floating-point-suffix}} & \rhdr{type} \\ \capsep
+none & \keyword{double} \\
+\tcode{f} or \tcode{F} & \keyword {float} \\
+\tcode{l} or \tcode{L} & \keyword{long} \keyword{double} \\
+\end{simpletypetable}
+
+\pnum
+\indextext{literal!floating-point}%
+The \defn{significand} of a \grammarterm{floating-point-literal}
+is the \grammarterm{fractional-constant} or \grammarterm{digit-sequence}
+of a \grammarterm{decimal-floating-point-literal}
+or the \grammarterm{hexadecimal-fractional-constant}
+or \grammarterm{hexadecimal-digit-sequence}
+of a \grammarterm{hexadecimal-floating-point-literal}.
+A \grammarterm{digit-sequence} is interpreted as a decimal (base ten) number.
+A \grammarterm{hexadecimal-digit-sequence} is interpreted
+as a hexadecimal (base sixteen) number.
```

Comment:
I find this to be underspecified in the fractional case; the number 0123.0123 is not "one hundred and twenty three point one hundred and twenty three". Similar to what we did for the integer case, can we replace all of this with:

> In the significand, the sequence of *digit*s or *hexadecimal-digit*s and optional period are interpreted as a base $N$ real number, where $N$ is 10 for a *decimal-floating-point-literal* or 16 for a *hexadecimal-floating-point-literal*.

(maybe with a table) and get rid of the "optional separating single quotes [...] are ignored" wording below?

Reply:
A table with two lines? Seems overkill.
Used "and" not "or" for the 10 vs. 16 part (both parts apply, under their respective conditions).
I'd prefer to turn the single quotes sentence into a note, similar to lex.icon.

Reply:
Using a note sounds good to me. I'm fine with either a table or non-table approach (slight preference for non-table), but was leaving the option open in case you would favor consistency with the integer literals section.